src/NServiceBus.Core/Pipeline/Outgoing/RoutingToDispatchConnector.cs

```diff
@@ -17,30 +17,18 @@ public override Task Invoke(IRoutingContext context, Func<IDispatchContext, Task
             var state = context.Extensions.GetOrCreate<State>();
             var dispatchConsistency = state.ImmediateDispatch ? DispatchConsistency.Isolated : DispatchConsistency.Default;
 
-            var operations = context.RoutingStrategies
-                .Select(rs =>
-                {
-                    var addressLabel = rs.Apply(context.Message.Headers);
-                    var message = new OutgoingMessage(context.Message.MessageId, context.Message.Headers, context.Message.Body);
-                    return new TransportOperation(message, addressLabel, dispatchConsistency, context.Extensions.GetDeliveryConstraints());
-                }).ToArray();
-
-            if (log.IsDebugEnabled)
+            var operations = new TransportOperation[context.RoutingStrategies.Count];
+            var index = 0;
+            foreach (var strategy in context.RoutingStrategies)
             {
-                var sb = new StringBuilder();
-                foreach (var operation in operations)
-                {
-                    var unicastAddressTag = operation.AddressTag as UnicastAddressTag;
-                    if (unicastAddressTag != null)
-                    {
-                        sb.AppendFormat("Destination: {0}\n", unicastAddressTag.Destination);
-                    }
-
-                    sb.AppendFormat("Message headers:\n{0}", string.Join(", ", operation.Message.Headers.Select(h => h.Key + ":" + h.Value).ToArray()));
-                    log.Debug(sb.ToString());
-                }
+                var addressLabel = strategy.Apply(context.Message.Headers);
+                var message = new OutgoingMessage(context.Message.MessageId, context.Message.Headers, context.Message.Body);
+                operations[index] = new TransportOperation(message, addressLabel, dispatchConsistency, context.Extensions.GetDeliveryConstraints());
+                index++;
             }
 
+            LogOutoingOperationsWhenDebugEnabled(operations);
```

Comment:
This would be better as
```csharp
if (isDebugEnabled)
{
    LogOutoingOperations(operations);
}
```
if you make `isDebugEnabled` declared as `static readonly bool`. That way `isDebugEnabled` will be a JIT-time constant, and the entire check and method call will be elided when the JIT compiles the method when debug isn't enabled.